### ISSUE TITLE
Build Error: illegal text reloc in '__decimal_exec' (from Modules/_decimal/_decimal.o) to '_mpd_round_string' for architecture arm64

### ISSUE THREAD

**Issue by furkanonder** [python/cpython#107452]
Build command;
```bash
./configure --with-pydebug && make -j
```

# Build Environment
- Operating system and architecture: macOS 13.4.1 22F82 arm64 

```bash
sh-3.2$ uname -a
Darwin arfs-MacBook-Pro.local 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:19 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T8103 arm64
sh-3.2$ gcc -v
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Output:

```bash
gcc -bundle -undefined dynamic_lookup      Modules/pyexpat.o -lm Modules/expat/libexpat.a  -o Modules/pyexpat.cpython-313d-darwin.so
ld: warning: ignoring file Modules/expat/libexpat.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
gcc -bundle -undefined dynamic_lookup      Modules/_pickle.o   -o Modules/_pickle.cpython-313d-darwin.so
gcc -bundle -undefined dynamic_lookup      Modules/_elementtree.o   -o Modules/_elementtree.cpython-313d-darwin.so
gcc -bundle -undefined dynamic_lookup      Modules/_decimal/_decimal.o -lm Modules/_decimal/libmpdec/libmpdec.a  -o Modules/_decimal.cpython-313d-darwin.so
ld: warning: ignoring file Modules/_decimal/libmpdec/libmpdec.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
ld: illegal text reloc in '__decimal_exec' (from Modules/_decimal/_decimal.o) to '_mpd_round_string' for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Modules/_decimal.cpython-313d-darwin.so] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by ned-deily** [comment]
Thanks for the report. You didn't identify what Python source you are trying to build but it appears to be 3.13 so I assume it is the main branch of the cpython repo. I am not able to reproduce this failure on a current macOS 13.5 arm64 system with the current HEAD (5113ed7a2b92e8beabebe5fe2f6e856c52fbe1a0) of the cpython main (3.13) branch. Suggest you ensure your local git repo is up-to-date with the upstream version, do a `make clean` of your build directory and try again.  If the problem persist, please describe what you are trying to build.

**Comment by furkanonder** [comment]
I am trying to build cpython in the main branch . The fork seems up to date.

```bash
bash-3.2$ git remote -v
origin	https://github.com/furkanonder/cpython (fetch)
origin	https://github.com/furkanonder/cpython (push)
upstream	https://github.com/python/cpython (fetch)
upstream	git@github.com:furkanonder/cpython.git (push)
bash-3.2$ git pull upstream main
From https://github.com/python/cpython
 * branch                  main       -> FETCH_HEAD
Already up to date.
bash-3.2$ git checkout
Your branch is up to date with 'upstream/main'.
```

I get the same error:
```python
gcc -bundle -undefined dynamic_lookup      Modules/_decimal/_decimal.o -lm Modules/_decimal/libmpdec/libmpdec.a  -o Modules/_decimal.cpython-313d-darwin.so
ld: warning: ignoring file Modules/_decimal/libmpdec/libmpdec.a, building for macOS-arm64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 0x20 )
ld: illegal text reloc in '__decimal_exec' (from Modules/_decimal/_decimal.o) to '_mpd_round_string' for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Modules/_decimal.cpython-313d-darwin.so] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by ned-deily** [comment]
It looks like you are running into another variation of the problem described in #102533; the root cause there and likely in what you seeing is that the build is not using the Apple-supplied version of the `ar` utility; perhaps as in that issue, you have the Homebrew `binutils` package installed and that version is appearing on the shell `PATH` before the Apple one; a `which ar` should show that.

There is an easy workaround until the fix for #102533 is released: you can specify to the Python configure script which `ar` to use with the `AR` variable:

`./configure --with-pydebug AR=$(xcrun -find ar) && make -j
`

That should find the Apple-supplied version of `ar` in either Xcode or in the Command Line Tools.

**Comment by furkanonder** [comment]
The `ar` package appears in the `binutils` package on the shell.

```bash
bash-3.2$ which ar
/opt/homebrew/opt/binutils/bin/ar
```

I built cpython using the `ar` of Command Line Tools.

```bash
bash-3.2$ xcrun -find ar
/Library/Developer/CommandLineTools/usr/bin/ar
```

Thanks for your help!
